lib/hotel_controller.rb

```diff
@@ -0,0 +1,228 @@
+    # done already in wave 1
+  
+    def reserve_room(start_date, end_date)
+      given_date_range = Hotel::DateRange.new(start_date, end_date)
+      avialable_rooms_list = available_rooms(start_date, end_date)
+      if avialable_rooms_list.empty?
+        raise NoRoomAvailableError.new("there is no available room.")
+      end
+      id = reservations.length + 1      
+      reservation = Hotel::Reservation.new(id,given_date_range, avialable_rooms_list.first)
+      add_reservation(reservation)
+      return reservation
+    end  
+
+    ############## Wave 3
+
+    # Add hotel_block method
+    def add_hotel_block(hotel_block)
+      @hotel_blocks << hotel_block
+    end
+    # Get the hotel_block_list for a given date (not exact match date range)
+    def hotel_block_list(start_date, end_date)
+      given_date_range = Hotel::DateRange.new(start_date, end_date)
+      # get the list of the hotel_block for the given date
+      hotel_block_list = hotel_blocks.select do |hotel_block|
+        hotel_block.date_range.overlap?(given_date_range)
+      end
+      return hotel_block_list
+    end
+
+    # Get the rooms_list_for_hotel_block
+    def rooms_list_for_hotel_block(start_date, end_date) 
+      hotel_block_list = hotel_block_list(start_date, end_date)
+      # get all arrays of rooms from the hotel_block_list for the given date 
+      rooms_hotel_block_arrays = hotel_block_list.map do |hotel_block|
+        hotel_block.rooms
+      end
+      # get all the rooms from the list of 
+      all_rooms_hotel_block =  rooms_hotel_block_arrays.flatten
+      return all_rooms_hotel_block
+    end 
+    
+    # Create a hotel_block
+    def create_hotel_block(date_range, rooms_array, discount_rate)
+      # all rooms of rooms_array must be available (not in the reservations list) during the given date range
+      all_reserved_rooms = reserved_rooms_list(date_range.start_date, date_range.end_date)
+  
+      # Cannot create another hotel block that any existing block includes that specific room for that specific date
+      all_rooms_hotel_block = rooms_list_for_hotel_block(date_range.start_date, date_range.end_date) 
+      # an exception raised if I try to create a Hotel Block 
+      # and at least one of the rooms is unavailable for the given date range
+
+      rooms_array.each do |room|
+        if all_reserved_rooms.any? {|r|r == room} || all_rooms_hotel_block.any? {|r| r == room}
+          raise ArgumentError.new("There is reservation conflict")
+        end
+      end
+      hotel_block = Hotel::HotelBlock.new(date_range, rooms_array, discount_rate)
+      add_hotel_block(hotel_block)
+      return hotel_block
+    end
+
+    # check whether a given block has any rooms available
+    def available_rooms_of_block(hotel_block) 
+      if hotel_block.rooms.empty?
+        raise ArgumentError.new("The hotel block cannot be made without having a room")
+      else
+        return hotel_block.rooms
+      end
+    end
```

Comment:
This method also seems like something the `HotelBlock` should be in charge of. if it were in the `HotelBlock`, it might look like this
```suggestion
    def available_rooms
      if self.rooms.empty?
        raise ArgumentError.new("The hotel block cannot be made without having a room")
      else
        return rooms
      end
    end
```

Moving this code both makes `HotelController` easier to read, and makes for fewer chained method calls.